### ISSUE TITLE
record.merge() getting slower as data grows in H2 database 

### ISSUE THREAD

**Issue by kran** [jOOQ/jOOQ#15590]
### Expected behavior

Hopefully performance won't drop so fast

### Actual behavior

As the amount of data increases linearly, the insertion speed becomes slower and slower. When it reaches 40,000 data, the performance is almost unacceptable.

### Steps to reproduce the problem

this is the table:

```sql
CREATE CACHED TABLE "PUBLIC"."VOD_RES"(
    "VOD_ID" INTEGER GENERATED BY DEFAULT AS IDENTITY(START WITH 1 RESTART WITH 46463) DEFAULT ON NULL NOT NULL,
    "VOD_OUT_ID" CHARACTER VARYING(20),
    "VOD_NAME" CHARACTER VARYING(255) DEFAULT '' NOT NULL,
    "VOD_SUB" CHARACTER VARYING(255) DEFAULT '' NOT NULL,
    "VOD_TAG" CHARACTER VARYING(100) DEFAULT '' NOT NULL,
    "VOD_CLASS" CHARACTER VARYING(255) DEFAULT '' NOT NULL,
    "VOD_PIC" CHARACTER VARYING(1024) DEFAULT '' NOT NULL,
    "VOD_ACTOR" CHARACTER VARYING(255) DEFAULT '' NOT NULL,
    "VOD_DIRECTOR" CHARACTER VARYING(255) DEFAULT '' NOT NULL,
    "VOD_WRITER" CHARACTER VARYING(100) DEFAULT '' NOT NULL,
    "VOD_BLURB" CHARACTER VARYING(255) DEFAULT '' NOT NULL,
    "VOD_REMARKS" CHARACTER VARYING(100) DEFAULT '' NOT NULL,
    "VOD_PUBDATE" CHARACTER VARYING(100) DEFAULT '' NOT NULL,
    "VOD_TOTAL" INTEGER DEFAULT '0' NOT NULL,
    "VOD_AREA" CHARACTER VARYING(20) DEFAULT '' NOT NULL,
    "VOD_LANG" CHARACTER VARYING(10) DEFAULT '' NOT NULL,
    "VOD_YEAR" CHARACTER VARYING(10) DEFAULT '' NOT NULL,
    "VOD_STATE" CHARACTER VARYING(30) DEFAULT '' NOT NULL,
    "VOD_ISEND" TINYINT DEFAULT '0' NOT NULL,
    "VOD_TIME" TIMESTAMP,
    "VOD_TIME_ADD" INTEGER DEFAULT '0' NOT NULL,
    "VOD_DOUBAN_ID" INTEGER DEFAULT '0' NOT NULL,
    "VOD_DOUBAN_SCORE" DECIMAL(3, 1) DEFAULT '0.0' NOT NULL,
    "VOD_PLAY_FROM" CHARACTER VARYING(255) DEFAULT '' NOT NULL,
    "VOD_PLAY_URL" CHARACTER VARYING NOT NULL,
    "VOD_DOWN_FROM" CHARACTER VARYING(255) DEFAULT '' NOT NULL,
    "VOD_DOWN_URL" CHARACTER VARYING NOT NULL,
    "TYPE_NAME" CHARACTER VARYING(20) DEFAULT NULL,
    "VID" INTEGER DEFAULT NULL,
    "DISTANCE" REAL DEFAULT 1,
    "DISTANCE_DATA" CHARACTER VARYING(200) DEFAULT NULL
);

ALTER TABLE "PUBLIC"."VOD_RES" ADD CONSTRAINT "PUBLIC"."CONSTRAINT_501" PRIMARY KEY("VOD_ID");

ALTER TABLE "PUBLIC"."VOD_RES" ADD CONSTRAINT "PUBLIC"."CONSTRAINT_5011" UNIQUE("VOD_OUT_ID");
```

just loop to insert rows with `record.merge()`.

### jOOQ Version

3.18.4

### Database product and version

H2

### Java Version

jdk21

### OS Version

macos ventura 13.3.1 (22E261)

### JDBC driver name and version (include name if unofficial driver)

org.h2.Driver

**Comment by lukaseder** [comment]
Thanks for your message.

Do you have any indication that jOOQ is doing anything wrong here? I would imagine this is more likely due to some bad plan within H2. You can try to profile your application e.g. using the free [VisualVM](https://visualvm.github.io/) profiler.

The `UpdatableRecord.merge()` call has to check for matches on all known unique constraints using an `OR` predicate, so perhaps that is something H2 can't handle really well from within a `MERGE` statement.

I don't think jOOQ can do much here, unless you have a specific suggestion?

**Comment by katzyn** [comment]
If you use H2 2.2.222, you need to run the `ANALYZE` command after insertion of enough number of rows, this version has a bug that prevents automatic collection of statistics; other versions aren't affected.

**Comment by lukaseder** [comment]
Thanks a lot, @katzyn !

**Comment by kran** [comment]
> If you use H2 2.2.222, you need to run the `ANALYZE` command after insertion of enough number of rows, this version has a bug that prevents automatic collection of statistics; other versions aren't affected.

@katzyn yes, i use h2 2.2.222, after found problem with `merge`, then i change it to insert, no analyze called and it works very well now. 

if the bug only affect merge operation?

**Comment by katzyn** [comment]
It affects performance of some queries. `INSERT` shoudn't be affected, because it doesn't use them internally.

This issue was fixed in H2 2.2.224.
